File: week-4/generate-parentheses/generate-parentheses.py
class Solution:
    def generateParenthesis(self, n: int) -> List[str]:
        res = []
        self.backtrack(0, 0, '', n, res)
        return res
    
    def backtrack(self, left, right, cur_string, n, res):
        if len(cur_string) == n*2:
            res.append(cur_string)
            return
        
        if left < n: self.backtrack(left + 1, right, cur_string + '(', n, res)
        if right < left: self.backtrack(left, right + 1, cur_string + ')', n, res)

File: week-4/generate-parentheses/generate-parentheses.js
/**
 * @param {number} n
 * @return {string[]}
 */
var generateParenthesis = function(n) {
    const res = [];
    backtrack(0, 0, '', n, res);
    return res;
};

const backtrack = (left, right, cur_string, n, res) => {
    if (cur_string.length === n*2){
        res.push(cur_string);
        return;
    }
    
    if (left<n) backtrack(left + 1, right, cur_string + '(', n, res);
    if (right<left) backtrack(left, right + 1, cur_string + ')', n, res);
};
